src/libmusig2.c

```diff
@@ -245,22 +254,53 @@ int musig2_aggregate_pubkey(musig2_context *mc, secp256k1_pubkey *pk_list, int n
     return 1;
 }
 
-int musig2_aggregate_R(musig2_context *mc, secp256k1_pubkey batch_list[][V]) {
+int musig2_aggregate_R(musig2_context *mc, secp256k1_pubkey batch_list[][V], int state) {
 
     int i, j;
     secp256k1_pubkey* temp_R_list[mc->nr_signers];
 
     /* Aggregate the batch commitments for current message */
     for (j = 0; j < V; j++) {
+        mc->aggr_R_list[state * V + j] = malloc(sizeof (secp256k1_pubkey));
         for (i = 0; i < mc->nr_signers; i++)
             temp_R_list[i] = &batch_list[i][j];
 
-        if (!secp256k1_ec_pubkey_combine(mc->ctx, &mc->aggr_R_list[j], (const secp256k1_pubkey *const *) temp_R_list, mc->nr_signers))
+        if (!secp256k1_ec_pubkey_combine(mc->ctx, mc->aggr_R_list[state * V + j], (const secp256k1_pubkey *const *) temp_R_list, mc->nr_signers))
             return 0;
     }
     return 1;
 }
 
+int musig2_signer_precomputation(musig2_context *mc, secp256k1_pubkey *pk_list, unsigned char *serialized_batch_list, int nr_signers, int nr_messages){
+    int i, j, k, ind;
+    int cnt = 0;
+    mc->nr_signers = nr_signers;
+    mc->aggr_R_list = malloc(sizeof (secp256k1_pubkey*) * nr_messages * V);
+    secp256k1_pubkey batch_list[nr_messages][nr_signers][V];   // Stores the batches of signers
+
+    // Parse the batch commitments of the signers
+    for (i = 0; i < nr_signers; i++) {
+        for (k = 0; k < nr_messages; k++) {
+            for (j = 0; j < V; j++) {
+                ind = (k * nr_signers * V + i * V + j) * SER_PK_BYTES;
+                assert(secp256k1_ec_pubkey_parse(mc->ctx, &batch_list[k][i][j], &serialized_batch_list[ind], SER_PK_BYTES));
+            }
+        }
+    }
+
+    // Aggregate R for each message to be signed.
+    for (k = 0; k < nr_messages; k++)
+        cnt += musig2_aggregate_R(mc, batch_list[k], k);
+
+    if (cnt != nr_messages)
+        return 0;
```

Comment:
Why do the aggregate for each index `k` and then fail if not _all_ worked, instead of checking the result at each iteration, and fail directly? The latter is more efficient in case of an early failure.